### ISSUE TITLE
Jupyter notebook filter to only spell check inside cell inputs

### ISSUE THREAD

**Issue by psychemedia** [codespell-project/codespell#2138]
The Jupyter notebook `.ipynb` document format combines markdown text, code, and code outputs (embedded images, data tables) inside a JSON document structure.

Spelling checking notebooks with cleared output cells generally works fine, but in some workflows it may be more convenient to spell check run notebooks. However, testing cell outputs may generate a large number of essentially false positive spelling errors.

It would be useful to be able to invoke a `codespell` switch (for example, `--ipynb`) that would ensure that only cell inputs are spell checked. (The [`nbformat`](https://nbformat.readthedocs.io/en/latest/format_description.html#cell-types) package provides a parser that can be used to parse notebooks.)

Notebook cells may also include cell tag metadata. It might also be useful to be able to specify cell tags to control the spell checking at both cell input and output level. For example:

- `codespell-ignore`: ignore this cell;
- `codespell-check-output`: also check the output of this (code) cell.

**Comment by matkoniecz** [comment]
> However, testing cell outputs may generate a large number of essentially false positive spelling errors

For example? In which situation there will be something detectable as misspelling? Is it about binary data?

**Comment by peternewman** [comment]
> For example? In which situation there will be something detectable as misspelling? Is it about binary data?

Not from the looks of things, it seems to be a way to document and run example code, so presumably the documentation, the code and the inputs to the code could all have typos.

I wonder if this is better written into the nbformat package as an extension, with some sort of API provided by codespell, rather than expecting codespell to import and handle the limitless number of document formats.

**Comment by matkoniecz** [comment]
> presumably the documentation, the code and the inputs to the code could all have typos

But in such case it would be beneficial to catch them with codespell, right?

If I would have example in documentation then except rare cases I would find typos also there.

**Comment by peternewman** [comment]
Yes, but I think the point was there might be stuff you don't want to check (e.g. actual hex). I don't really know anything about Jupyter notebook other than 2 minutes on their website (and that they can't spell Jupiter... :laughing: ).

**Comment by psychemedia** [comment]
The Jupyter notebook `.ipynb` file type is JSON document type with a cell based structure (https://nbformat.readthedocs.io/en/latest/format_description.html#cell-types ).

Mardown cells contain markdown source that is rendered as HTML by the notebook UI. Code cells have a cell input that contains source code, and cell outputs that contain content relating to a rendering of an object returned from the last line of the executed code, including data tables, object descriptions, images, embedded audio or video files, HTML files etc.

Notebooks can be saved in a state where all the cell outputs are cleared, so you don't get any false positives  from codespell finding hashes in megabytes of raw image output. But sometimes it's more convenient to be able to run a spell checker over a notebook that does include cell outputs. In which case, it would just be much cleaner to run `codespell` over content you know is code or markdown (the code and markdown cell inputs) and not any other cruft that happens to be in the `.ipynb` file.

**Comment by psychemedia** [comment]
Re: "can't spell Jupiter" - it's a portmanteau: `Ju-Py-te-R`, representing the three language kernels supported by the original IPython notebooks when the project was renamed. (There are now [kernels](https://github.com/jupyter/jupyter/wiki/Jupyter-kernels) for many languages folk might want to use.)

**Comment by peternewman** [comment]
Thanks for the background @psychemedia . Although personally I don't think my opinion of how best to implement this has changed:

> I wonder if this is better written into the nbformat package as an extension, with some sort of API provided by codespell, rather than expecting codespell to import and handle the limitless number of document formats.

There is a near-infinite number of formats which may want special treatment (such as #2135). Pushing them all into the core codespell means more stuff to support when people might not have the experience of, let alone use, those tools, as well as providing lots of bloat or dependencies when people just want to spellcheck a plain text file.

**Comment by psychemedia** [comment]
@peternewman Understood. Is there an example anywhere of writing an extension for `codespell`?

**Comment by peternewman** [comment]
> @peternewman Understood. Is there an example anywhere of writing an extension for `codespell`?

Not currently, as such a thing doesn't currently exist.

I pondered a bit more in here:
https://github.com/codespell-project/codespell/issues/2135#issuecomment-968090796

Perhaps start with a little tool that can iterate all the cells in a notebook via nbformat and output their text? As a hacky version you could just pass that text into codespell via STDIN and go from there (depending on how much you want a solution to the problem versus a full tool...

Also what would you want to do about config in a standalone tool?

The code which actually does the checking is here, I guess we need to turn that into a standalone function.
https://github.com/codespell-project/codespell/blob/99f39bd35950df54cf0cf09149654aef49da4107/codespell_lib/_codespell.py#L640-L712

**Comment by peternewman** [comment]
@psychemedia , I wonder if I've found an even easier way to integrate the two while allowing cross-language working...

See:
https://github.com/codespell-project/codespell/issues/1147#issuecomment-984204144

**Comment by psychemedia** [comment]
Ah, insteresting.. so just extract comments and then spell check those. I think my original motivation was to be able to make use of the Jupyter notebook structure when trying to spell check `.ipynb` docs.

The approach I use now is to just to clear cell outputs and then spell check the notebook as is. An alternative approach is to convert `.ipynb` to a text format (eg markdown or `.py`) using `jupytext` and then spell check that. (That gets rid of all the cell outputs as well as the JSON structure in the `.ipynb`.) Your new suggested approach could improve that text pipeline route: `.ipynb` -> `.py` -> extract comments -> `codespell`

**Comment by peternewman** [comment]
> Ah, insteresting.. so just extract comments and then spell check those.

I'd think of it more as remove stuff that isn't comments rather than extract comments. If you're able to just strip the stuff that will trip up codespell, but keep the comments in exactly the same places, then with some fairly minor fudging of codespell or it's output, my main draw was you could use the annotation info coming out as-is, i.e. if codespell said it was on line 5, it would be in exactly that place in the original notebook.

> I think my original motivation was to be able to make use of the Jupyter notebook structure when trying to spell check `.ipynb` docs.

I don't really follow this, would it provide something more human friendly to edit within the notebook itself, like example one question section (or whatever is relevant to these)?

> The approach I use now is to just to clear cell outputs and then spell check the notebook as is. An alternative approach is to convert `.ipynb` to a text format (eg markdown or `.py`) using `jupytext` and then spell check that. (That gets rid of all the cell outputs as well as the JSON structure in the `.ipynb`.) Your new suggested approach could improve that text pipeline route: `.ipynb` -> `.py` -> extract comments -> `codespell`

In the short term, if you write something that takes a .ipynb file as input on stdin and returns the cleaned version on stdout, then for now you can do something like:
`cat foo.ipynb | python3 filter.py | codespell -`

It will say:
speling ==> spelling

And you'll get some useful output, and then with fairly minimal changes to codespell we could let you do something like:
`codespell --filters="ipynb=python3 filter.py" foo.ipynb`

And it would instead say:
foo.ipynb:42: speling ==> spelling

Which obviously doesn't make much difference for a single file, but would be more important when you're checking a whole folder of them!

**Comment by psychemedia** [comment]
Re: "more human friendly", this is the view I tend to edit files in eg in VS Code:

![](https://opencomputinglab.github.io/OpenJALE/images/vscode.png0

or in Jupyter notebook:

![image](https://user-images.githubusercontent.com/82988/144853148-544a5c33-dd74-4c7f-9dea-2adf7089d147.png)

Using the [`jupytext`](https://jupytext.readthedocs.io/en/latest/) extension, I can edit [`light`](https://jupytext.readthedocs.io/en/latest/formats.html#the-light-format) or [`percent`](https://jupytext.readthedocs.io/en/latest/formats.html#the-percent-format) format `.py`files in a notebook UI - so for example, comments can be styled in markdown cells, and code split out into code cells.

**Comment by psychemedia** [comment]
Re: the pipeline - yep, one appraoch would be to do s/thing like:

`cat $MY_NB_PATH | jupytext --from ipynb --to py:percent | codespell -`

Though the line numbers aren't totally meanigful without access to the (dynamically created) `.py` file.

**Comment by EwoutH** [comment]
> I also have an issue where Jupyter Notebook images become corrupted due to codespell. In [this commit](https://github.com/quaquel/EMAworkbench/pull/195/commits/783dacbf759b3b7d6d0e9e0bc3eb2cb25527e6a6) serveral images are modified and breaking when using Codespell v2.2.1.
> 
> A solution where codespell ignores content in `"data": {"image/*"}` fields by default would be ideal.

Unfortunately, this is still the case and I keep encountering this bug once every while.

- https://github.com/projectmesa/mesa/pull/2593

@DimitriPapadopoulos would you mind taking a look at this issue?

**Comment by DimitriPapadopoulos** [comment]
I don't think this can be easily fixed, because codespell is programming language-agnostic. See #17, #1147, #2790 for similar issues.
